vault/identity_store_oidc_provider.go

```diff
@@ -345,3 +445,129 @@ func (i *IdentityStore) pathOIDCScopeExistenceCheck(ctx context.Context, req *lo
 
 	return entry != nil, nil
 }
+
+// pathOIDCCreateUpdateProvider is used to create a new named provider or update an existing one
+func (i *IdentityStore) pathOIDCCreateUpdateProvider(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	name := d.Get("name").(string)
+
+	var provider provider
+	if req.Operation == logical.UpdateOperation {
+		entry, err := req.Storage.Get(ctx, providerPath+name)
+		if err != nil {
+			return nil, err
+		}
+		if entry != nil {
+			if err := entry.DecodeJSON(&provider); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	if issuerRaw, ok := d.GetOk("issuer"); ok {
+		provider.Issuer = issuerRaw.(string)
+	} else if req.Operation == logical.CreateOperation {
+		provider.Issuer = d.GetDefaultOrZero("issuer").(string)
+	}
+
+	if allowedClientIDsRaw, ok := d.GetOk("allowed_client_ids"); ok {
+		provider.AllowedClientIDs = allowedClientIDsRaw.([]string)
+	} else if req.Operation == logical.CreateOperation {
+		provider.AllowedClientIDs = d.GetDefaultOrZero("allowed_client_ids").([]string)
+	}
+
+	if scopesRaw, ok := d.GetOk("scopes"); ok {
+		provider.Scopes = scopesRaw.([]string)
+	} else if req.Operation == logical.CreateOperation {
+		provider.Scopes = d.GetDefaultOrZero("scopes").([]string)
+	}
+
+	scopeTemplateKeyNames := make(map[string]string)
+	for _, scopeName := range provider.Scopes {
+		entry, err := req.Storage.Get(ctx, scopePath+scopeName)
+		if err != nil {
+			return nil, err
+		}
+		// enforce scope existence on provider creation
+		if entry == nil {
+			return logical.ErrorResponse("cannot find scope %s", scopeName), nil
```

Comment:
nit: Maybe we should change this to the following for consistency, since the [Client PR was updated](https://github.com/hashicorp/vault/pull/12272#discussion_r692451524):

```suggestion
			return logical.ErrorResponse("scope %q does not exist", scopeName), nil
```